File: Sources/MoonRuntime/Executor.swift
import MoonAST
import MoonMoonfile
import MoonPlanner

func executeDag(_ dag: ExecutionDag, _ block: DoBlock, _ ctx: RuntimeContext) async throws {
    var completed = Set<String>()
    var pending = Set(dag.nodes.map(\.id))

    while !pending.isEmpty {
        let ready = dag.nodes.filter { node in
            pending.contains(node.id) && node.dependencies.allSatisfy { completed.contains($0) }
        }

        if ready.isEmpty {
            throw RuntimeError("DAG deadlock — cyclic dependencies")
        }

        try await withThrowingTaskGroup(of: String.self) { group in
            for node in ready {
                group.addTask {
                    try await executeDagNode(node, dag, block, ctx)
                    return node.id
                }
            }

            for try await nodeId in group {
                pending.remove(nodeId)
                completed.insert(nodeId)
            }
        }
    }

    ctx.env["_result"] = .null
}

private func executeDagNode(
    _ node: DagNode,
    _ dag: ExecutionDag,
    _ block: DoBlock,
    _ ctx: RuntimeContext
) async throws {
    let stmt = block.statements[node.stmtIndex]

    if node.kind == .mapM, case .bind(_, let expr, _, _) = stmt {
        if let mapM = detectMapM(expr), let listVar = node.mapMListVar {
            let list = ctx.env[listVar] ?? .array([])
            let fn = try await evalExpr(mapM.funcExpr, ctx)
            let tier = tierForMapM(node.mapMFunc ?? "", ctx)

            let results: [RuntimeValue]
            if case .array(let items) = list {
                results = try await ctx.pool.runAll(tier, items.map { item in
                    { try await applyFn(fn, item, ctx) }
                })
            } else {
                results = []
            }

            ctx.env["__mapM_\(node.id)"] = .array(results)
            return
        }
    }

    if node.kind == .mapM_join {
        let parent = dag.nodes.first { $0.kind == .mapM && node.dependencies.contains($0.id) }
        let results = parent.flatMap { ctx.env["__mapM_\($0.id)"] } ?? .array([])
        if let bindVar = node.bindVar {
            ctx.env[bindVar] = results
        }
        return
    }

    switch stmt {
    case .storm:
        let value = try await runStormStmt(stmt, ctx)
        if let bindVar = node.bindVar {
            ctx.env[bindVar] = value
        }
    case .bind(let pattern, let expr, let config, _):
        var value = try await evalExprWithConfig(expr, ctx, config: config)
        value = try await applyBindConfig(value, config, ctx)
        if let bindVar = node.bindVar {
            ctx.env[bindVar] = value
        } else {
            bindPattern(pattern, value, ctx)
        }
    default:
        _ = try await runDoStmt(stmt, ctx)
    }
}

private func tierForMapM(_ mapMFunc: String, _ ctx: RuntimeContext) -> ModelTier {
    let candidates = mapMFunc.split(whereSeparator: { !$0.isLetter && !$0.isNumber }).map(String.init)
    for name in candidates where name.first?.isUppercase == true {
        if let agent = ctx.agents[name] {
            return modelToTier(agentModel(agent))
        }
    }
    return .pro
}

func applyFn(_ fn: RuntimeValue, _ arg: RuntimeValue, _ ctx: RuntimeContext) async throws -> RuntimeValue {
    if case .callable(let box) = fn {
        return try await box.fn(arg, ctx)
    }
    return try await applyValue(fn, arg, ctx)
}

public struct ProgramRunOptions: Sendable {
    public var functionName: String
    public var entryPath: String?
    public var projectRoot: String?
    public var overrides: RuntimeConfigOverrides
    public var llm: LlmClient?
    public var workerPool: WorkerPool?
    public var traceLlm: Bool
    public var traceDir: String?

    public init(
        functionName: String = "main",
        entryPath: String? = nil,
        projectRoot: String? = nil,
        overrides: RuntimeConfigOverrides = RuntimeConfigOverrides(),
        llm: LlmClient? = nil,
        workerPool: WorkerPool? = nil,
        traceLlm: Bool = false,
        traceDir: String? = nil
    ) {
        self.functionName = functionName
        self.entryPath = entryPath
        self.projectRoot = projectRoot
        self.overrides = overrides
        self.llm = llm
        self.workerPool = workerPool
        self.traceLlm = traceLlm
        self.traceDir = traceDir
    }
}

public func runProgram(_ program: Program, options: ProgramRunOptions = ProgramRunOptions()) async throws -> ProgramRunResult {
    let functionName = options.functionName
    guard let dag = planFunction(program, functionName: functionName) else {
        throw RuntimeError("Function not found: \(functionName)")
    }
    guard let block = findFunction(program, name: functionName) else {
        throw RuntimeError("Function body not found: \(functionName)")
    }

    var mergedOverrides = options.overrides
    if let projectRoot = options.projectRoot,
       let moonfilePath = findMoonfile(startDir: projectRoot),
       let moonfile = try? loadMoonfile(path: moonfilePath) {
        mergedOverrides = mergeRuntimeOverrides(
            runtimeOverrides(from: moonfileToRuntimeOverrides(moonfile)),
            mergedOverrides
        )
    }
    let runtimeConfig = loadRuntimeConfig(overrides: mergedOverrides)
    configureTokenizer(path: runtimeConfig.tokenizerPath)
    let metrics = MetricsCollector(pricing: runtimeConfig.pricing)

    var llm: LlmClient = options.llm ?? createLlmClient(config: runtimeConfig, metrics: metrics)
    if options.traceLlm {
        let writer = try await LlmTraceWriter.create(baseDir: options.traceDir)
        llm = TracingLlmClient(client: llm, writer: writer)
    }

    let pool = options.workerPool ?? WorkerPool(config: WorkerPoolConfig(
        flashConcurrency: runtimeConfig.flashConcurrency,
        proConcurrency: runtimeConfig.proConcurrency,
        onAcquire: { tier, active in metrics.recordWorkerStart(tier: tier, concurrent: active) }
    ))

    let ctx = RuntimeContext(
        program: program,
        agents: collectAgents(program),
        builtins: builtinsFromImports(program),
        llm: llm,
        memory: MemoryManager(longTermPath: runtimeConfig.longTermMemoryPath, metrics: metrics),
        pool: pool,
        metrics: metrics,
        schemas: compiledSchemasForRuntime(program),
        systemSuffix: runtimeConfig.systemSuffix
    )

    try await executeDag(dag, block, ctx)

    return ProgramRunResult(
        value: ctx.env["_result"] ?? .null,
        effects: ctx.effects,
        dag: dag,
        metrics: metrics.snapshot()
    )
}

File: legacy/packages/runtime/src/executor.ts
import type { DoBlock, Program } from "@moon/ast";
import { resolveImports } from "@moon/resolver";
import { dirname, resolve } from "path";
import { builtinsFromImports } from "./stdlib/registry";
import type { ExecutionDag } from "@moon/planner";
import { detectMapM, findFunction, planFunction } from "@moon/planner";
import { compileSchemas } from "@moon/schema-compiler";
import { agentModel, collectAgents } from "./agents";
import {
  applyBindConfig,
  applyValue,
  defaultBuiltins,
  evalExpr,
  evalExprWithConfig,
  runDoStmt,
  type RuntimeContext,
  type RuntimeEffect,
  type RuntimeValue,
} from "./interpreter";
import type { LlmClient } from "./llm-types";
import { MetricsCollector, type RunMetrics } from "./metrics";
import { createLlmClient, loadRuntimeConfig, type RuntimeConfigOverrides } from "./runtime-config";
import { MemoryManager } from "./memory";
import { modelToTier, WorkerPool, type ModelTier } from "./worker-pool";
import { LlmTraceWriter, type LlmTraceSession } from "./llm-trace";
import { wrapLlmWithTrace } from "./tracing-llm";

export interface RunOptions extends RuntimeConfigOverrides {
  functionName?: string;
  entryPath?: string;
  projectRoot?: string;
  llm?: LlmClient;
  workerPool?: WorkerPool;
  traceLlm?: boolean;
  traceDir?: string;
}

export interface RunResult {
  value: RuntimeValue;
  effects: RuntimeEffect[];
  dag: ExecutionDag;
  metrics: RunMetrics;
  llmTrace?: LlmTraceSession;
}

export async function runProgram(program: Program, options: RunOptions = {}): Promise<RunResult> {
  const functionName = options.functionName ?? "main";
  const dag = planFunction(program, functionName);
  if (!dag) {
    throw new Error(`Function not found: ${functionName}`);
  }

  const block = findFunction(program, functionName);
  if (!block) {
    throw new Error(`Function body not found: ${functionName}`);
  }

  const runtimeConfig = loadRuntimeConfig(options);
  const metrics = new MetricsCollector(runtimeConfig.pricing);
  const agents = collectAgents(program);

  let traceWriter: LlmTraceWriter | undefined;
  let llm: LlmClient = options.llm ?? createLlmClient(runtimeConfig, metrics);
  if (options.traceLlm) {
    traceWriter = await LlmTraceWriter.create(options.traceDir);
    llm = wrapLlmWithTrace(llm, traceWriter);
  }

  const entryPath = options.entryPath;
  const resolved = entryPath
    ? resolveImports(program, {
        entryPath: resolve(entryPath),
        projectRoot: options.projectRoot ?? dirname(resolve(entryPath)),
      })
    : null;
  const builtins = resolved
    ? builtinsFromImports(resolved.imports)
    : defaultBuiltins();

  const ctx: RuntimeContext = {
    program,
    schemas: compileSchemas(program).schemas,
    agents,
    llm,
    memory: new MemoryManager({
      longTermPath: runtimeConfig.longTermMemoryPath,
      metrics,
    }),
    pool: options.workerPool ?? new WorkerPool({
      flashConcurrency: runtimeConfig.flashConcurrency,
      proConcurrency: runtimeConfig.proConcurrency,
      onAcquire: (tier, active) => metrics.recordWorkerStart(tier, active),
    }),
    env: new Map(),
    effects: [],
    builtins,
  };

  await executeDag(dag, block, ctx);
  return {
    value: ctx.env.get("_result") ?? null,
    effects: ctx.effects,
    dag,
    metrics: metrics.snapshot(),
    llmTrace: traceWriter?.snapshot(),
  };
}

async function executeDag(
  dag: ExecutionDag,
  block: DoBlock,
  ctx: RuntimeContext,
): Promise<void> {
  const completed = new Set<string>();
  const pending = new Set(dag.nodes.map((n) => n.id));

  while (pending.size > 0) {
    const ready = dag.nodes.filter(
      (n) => pending.has(n.id) && n.dependencies.every((d) => completed.has(d)),
    );

    if (ready.length === 0) {
      throw new Error("DAG deadlock — cyclic dependencies");
    }

    await Promise.all(
      ready.map(async (node) => {
        const stmt = block.statements[node.stmtIndex];
        if (!stmt) return;

        if (node.kind === "mapM" && stmt.kind === "Bind") {
          const mapM = detectMapM(stmt.expr);
          if (!mapM || !node.mapMListVar) return;

          const list = ctx.env.get(node.mapMListVar) as RuntimeValue[];
          const fn = await evalExpr(mapM.func, ctx);
          const tier = tierForMapM(node.mapMFunc ?? "", ctx);

          const results = await ctx.pool.runAll(
            tier,
            list.map((item) => async () => applyFn(fn, item, ctx)),
          );

          ctx.env.set(`__mapM_${node.id}`, results);
          pending.delete(node.id);
          completed.add(node.id);
          return;
        }

        if (node.kind === "mapM_join") {
          const parent = dag.nodes.find((n) => n.kind === "mapM" && node.dependencies.includes(n.id));
          const results = parent ? ctx.env.get(`__mapM_${parent.id}`) : [];
          if (node.bindVar) {
            ctx.env.set(node.bindVar, results);
          }
          pending.delete(node.id);
          completed.add(node.id);
          return;
        }

        if (stmt.kind === "Storm") {
          const { runStormStmt } = await import("./storm");
          const value = await runStormStmt(stmt, ctx);
          if (node.bindVar) ctx.env.set(node.bindVar, value);
        } else if (stmt.kind === "Bind") {
          let value = await evalExprWithConfig(stmt.expr, ctx, stmt.config);
          value = await applyBindConfig(value, stmt.config, ctx);
          if (node.bindVar) ctx.env.set(node.bindVar, value);
        } else {
          await runDoStmt(stmt, ctx);
        }

        pending.delete(node.id);
        completed.add(node.id);
      }),
    );
  }

  ctx.env.set("_result", null);
}

function tierForMapM(mapMFunc: string, ctx: RuntimeContext): ModelTier {
  const candidates = mapMFunc.match(/[A-Z][A-Za-z0-9]*/g) ?? [];
  for (const name of candidates) {
    const agent = ctx.agents.get(name);
    if (agent) {
      return modelToTier(agentModel(agent));
    }
  }
  return "pro";
}

async function applyFn(
  fn: RuntimeValue,
  arg: RuntimeValue,
  ctx: RuntimeContext,
): Promise<RuntimeValue> {
  if (typeof fn === "function") {
    const result = fn(arg);
    return result instanceof Promise ? result : result;
  }
  return applyValue(fn, arg, ctx);
}